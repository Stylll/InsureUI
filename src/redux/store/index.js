import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware from 'redux-saga';

import rootReducer from '../reducers';
import rootSaga from '../sagas/index';

const sagaMiddleware = createSagaMiddleware();
const middleware = composeWithDevTools(applyMiddleware(sagaMiddleware));
const store = createStore(rootReducer, middleware);

sagaMiddleware.run(rootSaga);

export default store;
