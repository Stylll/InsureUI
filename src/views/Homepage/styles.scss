@import '../../partials/colors';

.categories-container {
  display: flex;
  justify-content: center;

  .loader-container {
    margin-top: 50px;
    .message {
        font-size: 18px;
    }
  }

  .title {
    text-align: center;
  }

  .totalSum {
    text-align: left;
  }

  .categoryName {
    font-weight: 700;
  }

  .categoryTotal {
    text-align: right;
  }

  .noItemRecords {
    text-align: right;
  }

  table {
    border-collapse: collapse;
  }
    
  table, th, td {
      border: 1px solid $black;
  }

  th, td {
      padding: 15px;
      text-align: left;
  }

  tr:hover {
      background-color: $gray;
      cursor: pointer;
  }

  th {
      background-color: $primary-color;
      color: $white;
  }

  .deleteBtn {
    color: $red;
  }
}
