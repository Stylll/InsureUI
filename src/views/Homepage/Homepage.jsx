import React, { Fragment, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Header from '../../components/Header/Header.component';
import Loader from '../../components/Loader/Loader.component';
import Modal from '../../components/Modal/Modal.component';
import ConfirmPrompt from '../../components/ConfirmPrompt/ConfirmPrompt.component';
import * as categoryActions from '../../redux/actionCreators/categoryActions/categoryActions';
import * as itemActions from '../../redux/actionCreators/itemActions/itemActions';
import './styles.scss';

const Homepage = ({
  categoryItems, getCategoryItems, isLoading, isDeleteLoading, deleteItem,
  dataTestId,
}) => {
  const [showDeleteItemModal, setShowDeleteItemModal] = useState(false);
  const [itemToDeleteId, setItemToDeleteId] = useState(null);
  const [isDeletingItem, setIsDeletingItem] = useState(false);

  useEffect(() => {
    getCategoryItems();
  }, [getCategoryItems]);

  useEffect(() => {
    if (!isDeleteLoading && isDeletingItem) {
      hideDeleteItem();
    }
  }, [isDeleteLoading, isDeletingItem]);

  const loaderStyles = {
    width: '20px',
    height: '20px',
  };

  const showDeleteItem = (itemId) => {
    setItemToDeleteId(itemId);
    setShowDeleteItemModal(true);
  };

  const hideDeleteItem = () => {
    setItemToDeleteId(null);
    setShowDeleteItemModal(false);
    setIsDeletingItem(false);
  };

  const handleDeleteItem = () => {
    deleteItem(itemToDeleteId);
    setIsDeletingItem(true);
  };

  const renderCategoryItems = (categoryList) => {
    if (!categoryList || !categoryList.items.length) {
      return (
        <tr>
          <td className="noItemRecords" colSpan="4">No item records.</td>
        </tr>
      );
    }

    return (
      categoryList.items.map((item, key) => {
        const sn = key + 1;
        return (
          <tr key={key}>
            <td>{sn}</td>
            <td>{item.name}</td>
            <td>${item.value}</td>
            <td className="deleteBtn"
              onClick={() => { showDeleteItem(item.id); }}
              data-testid={`${dataTestId}-delete-${sn}`}
            >&times;</td>
          </tr>
        );
      })
    );
  };

  const renderItemsTotal = (total) => {
    if (total > 0) {
      return (
        <tr>
          <td className="categoryTotal" colSpan="4" align="right">Total: ${total}</td>
        </tr>
      );
    }
    return null;
  };

  const renderBody = () => {
    const { categories } = categoryItems;

    if (!categories || !categories.length) {
      return (
        <tbody>
        <tr>
          <td colSpan="4">No category records.</td>
        </tr>
      </tbody>
      );
    }

    return (
      <tbody>
            {categories.map((category, i) => (
                <Fragment key={i}>
                  <tr>
                    <td className="categoryName" colSpan="4">{category.name}</td>
                  </tr>
                  {renderCategoryItems(category)}
                  {renderItemsTotal(category.total)}
                </Fragment>
            ))}
          </tbody>
    );
  };

  const renderTable = () => {
    const { total = 0 } = categoryItems;
    if (!total) {
      return (
        <div>
          <h3>No Records Found.</h3>
        </div>
      );
    }
    return (
      <div className="table">
        <table>
          <thead>
            <tr>
              <th>S/N</th>
              <th>Item Name</th>
              <th>Value</th>
              <th />
            </tr>
          </thead>
          {renderBody()}
          <tfoot>
            <tr>
              <td className="totalSum" colSpan="4">Total Sum: ${total}</td>
            </tr>
          </tfoot>
        </table>
      </div>
    );
  };

  return (
    <div className="homepage-container">
      <Header />
      <div className="categories-container">
        <div>
          <h3 className="title">List Of Items</h3>
          {isLoading && <Loader
            customStyles={loaderStyles}
            message="fetching category items records, hang tight..." />}
          {!isLoading && renderTable()}
        </div>
      </div>
      <Modal isModalVisible={showDeleteItemModal}
        handleModalClose={hideDeleteItem}
      >
        <ConfirmPrompt
          handleCancel={hideDeleteItem}
          handleClick={handleDeleteItem}
          message="Are sure you want to delete this item?"
          isLoading={isDeleteLoading}
          />
      </Modal>
    </div>
  );
};

Homepage.defaultProps = {
  dataTestId: 'viewcategoryitems',
};

Homepage.propTypes = {
  getCategoryItems: PropTypes.func.isRequired,
  dataTestId: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  isDeleteLoading: PropTypes.bool.isRequired,
  getCategoryItemsSuccess: PropTypes.bool.isRequired,
  deleteItem: PropTypes.func.isRequired,
  errorMessage: PropTypes.string.isRequired,
  categoryItems: PropTypes.object.isRequired,
};

Homepage.defaultProps = {
  dataTestId: 'homepage',
};

const mapStateToProps = ({ category, item }) => ({
  isLoading: category.isCategoryItemsLoading,
  isDeleteLoading: item.isDeleteLoading,
  getCategoryItemsSuccess: category.getCategoryItemsSuccess,
  errorMessage: category.getCategoryItemsErrorMessage,
  categoryItems: category.categoryItems,
});

const mapDispatchToProps = {
  getCategoryItems: categoryActions.getCategoryItems,
  deleteItem: itemActions.deleteItem,
};

export default connect(mapStateToProps, mapDispatchToProps)(Homepage);
