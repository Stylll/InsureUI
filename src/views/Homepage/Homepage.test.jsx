import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import Homepage from './Homepage';
import store from '../../redux/store';
import {
  getCategoryItemsSuccess,
} from '../../redux/actionCreators/categoryActions/categoryActions';
import {
  getCategoryItemsApiResponse,
} from '../../redux/testhelpers';

describe('Homepage Test', () => {
  it('should render properly', () => {
    const { baseElement } = render(
      <Provider store={store}>
        <Homepage />
      </Provider>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should render fetching categories records', () => {
    const { getByText } = render(
        <Provider store={store}>
            <Homepage />
        </Provider>,
    );
    expect(getByText('fetching category items records, hang tight...')).toBeInTheDocument();
    store.dispatch(getCategoryItemsSuccess({}));
    expect(getByText('No Records Found.')).toBeInTheDocument();
  });

  it('should render list of accounts', () => {
    const { getByText } = render(
        <Provider store={store}>
            <Homepage />
        </Provider>,
    );
    store.dispatch(getCategoryItemsSuccess(getCategoryItemsApiResponse.data.data));
    expect(getByText('Electronics')).toBeInTheDocument();
    expect(getByText('Playstation')).toBeInTheDocument();
  });

  it('should render and close create item form', () => {
    const { getByTestId, getByText, queryByText } = render(
        <Provider store={store}>
            <Homepage />
        </Provider>,
    );
    fireEvent.click(getByTestId('add-item'));
    expect(getByText('Add an Item')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Add an Item')).not.toBeInTheDocument();
  });

  it('should render and close delete item confirm prompt', () => {
    const { getByTestId, getByText, queryByText } = render(
        <Provider store={store}>
            <Homepage />
        </Provider>,
    );
    store.dispatch(getCategoryItemsSuccess(getCategoryItemsApiResponse.data.data));
    fireEvent.click(getByTestId('homepage-delete-1'));
    expect(getByText('Are sure you want to delete this item?')).toBeInTheDocument();
    fireEvent.click(getByTestId('modal-close'));
    expect(queryByText('Are sure you want to delete this item?')).not.toBeInTheDocument();
  });
});
