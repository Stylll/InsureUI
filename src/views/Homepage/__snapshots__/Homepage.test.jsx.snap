// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Test should render properly 1`] = `
<body>
  <div>
    <div
      class="homepage-container"
    >
      <div>
        <div
          class="header container"
        >
          <div
            class="left"
          >
            <div
              class="link"
            >
              Insure App
            </div>
          </div>
          <div
            class="right"
          >
            <div
              class="link"
              data-testid="add-item"
            >
              Add Item
            </div>
          </div>
        </div>
        <div
          class="modal"
        >
          <div
            class="modal-container"
          >
            <div
              class="close-btn"
              data-testid="modal-close"
            >
              ×
            </div>
            <div
              class="content "
            >
              <div
                class="create-item-container"
              >
                <div
                  class="form-container"
                >
                  <div
                    class="item-container"
                  >
                    <div>
                      <p>
                        Add an Item
                      </p>
                      <span
                        class="error-text"
                      />
                      <div
                        class="form-input"
                      >
                        <div
                          class="title"
                        >
                          Enter the item name
                        </div>
                        <input
                          class="text-input"
                          data-testid="createitem-name"
                          id=""
                          name="Name"
                          placeholder="Playstation"
                          type="text"
                          value=""
                        />
                        <span
                          class="info"
                        />
                        
                      </div>
                      <div
                        class="form-input"
                      >
                        <div
                          class="title"
                        >
                          Enter the value of the item
                        </div>
                        <input
                          class="text-input"
                          data-testid="createitem-value"
                          id=""
                          name="Value"
                          placeholder="5000"
                          type="text"
                          value=""
                        />
                        <span
                          class="info"
                        />
                        
                      </div>
                      <div
                        class="drop-down"
                      >
                        <select
                          class="select"
                          data-testid="createitem-category"
                          name="CategoryId"
                        >
                          <option
                            value=""
                          >
                            select item category
                          </option>
                          <option
                            value="1"
                          >
                            Electronics
                          </option>
                          <option
                            value="2"
                          >
                            Clothing
                          </option>
                          <option
                            value="3"
                          >
                            Kitchen
                          </option>
                        </select>
                        <span
                          class="chevy"
                        >
                          ∨
                        </span>
                        <span
                          class="info"
                        />
                        
                      </div>
                      <br />
                      <div
                        class="btn-button"
                      >
                        <button
                          class="button "
                          data-testid="createitem-create"
                          type="submit"
                        >
                          Create
                        </button>
                      </div>
                    </div>
                    <br />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      Homepage
    </div>
  </div>
</body>
`;
